frontend/lib/resolution/Resolver.cpp

```diff
@@ -3920,6 +3906,53 @@ ResolvedExpression Resolver::resolveNameInModule(const UniqueString name) {
   return result;
 }
 
+static QualifiedType computeDefaultsIfNecessary(Resolver& rv,
+                                                const QualifiedType& type,
+                                                const ID& id,
+                                                const Identifier* ident) {
+  // now, for a type that is generic with defaults,
+  // compute the default version when needed. e.g.
+  //   record R { type t = int; }
+  //   var x: R; // should refer to R(int)
+  bool computeDefaults = true;
+  bool resolvingCalledIdent = rv.nearestCalledExpression() == ident;
+
+  // For calls like
+  //
+  //   type myType = anotherType(int)
+  //
+  // Use the generic version of anotherType to feed as receiver.
+  if (resolvingCalledIdent) {
+    computeDefaults = false;
+  }
+
+  // If we're referring to variable-ish thing, don't instantiate
+  // generics. This way, `type t = someGeneric(?); t` doesn't instantiate.
+  // Peformance: finding the parent tag is pretty expensive. Can we fold
+  // the knowledge into IdAndFlags?
+  if (id && asttags::isVarLikeDecl(parsing::idToTag(rv.context, id))) {
```

Comment:
This kind of situation, where we're checking if we're inside a particular kind of parent node, seems less than ideal to me. Could we make the decision about default-ness while actively handling those parent nodes, instead of looking upward? Or perhaps we've already considered and dismissed that for various reasons, I can't quite recall.

Reply:
I'd like to leave this as a follow-up because 1) the code predates this PR and 2) I don't know that we can get around this.

Also, the comment is wrong -- we're not finding the parent ID, but rather the AST. This is relatively expensive too. So, in fact, the parent-child relationship is not a given.

Given that, I can still see a re-formulation of your suggestion: determine whether a variable or type declaration should use defaults ahead of time, and don't bother special-casing when it's referenced. The problem is that whether we use defaults is context-sensitive anyway (see above code which checks for `nearestCalledIdent()`, and in some cases we do not want them, so we can't eagerly create defaulted types.